File: client-py/queen/stream/stream_builder.py
"""
StreamBuilder - Fluent API for defining streams
"""

from typing import Any, Dict, List


class StreamBuilder:
    """StreamBuilder - Fluent API for defining streams"""

    def __init__(self, http_client: Any, queen: Any, name: str, namespace: str):
        """
        Initialize stream builder

        Args:
            http_client: HttpClient instance
            queen: Queen instance
            name: Stream name
            namespace: Stream namespace
        """
        self._http_client = http_client
        self._queen = queen
        self._config: Dict[str, Any] = {
            "name": name,
            "namespace": namespace,
            "source_queue_names": [],
            "partitioned": False,
            "window_type": "tumbling",
            "window_duration_ms": 60000,
            "window_grace_period_ms": 30000,
            "window_lease_timeout_ms": 60000,
        }

    def sources(self, queue_names: List[str]) -> "StreamBuilder":
        """
        Set source queues

        Args:
            queue_names: Array of queue names

        Returns:
            Self for chaining
        """
        self._config["source_queue_names"] = queue_names
        return self

    def partitioned(self) -> "StreamBuilder":
        """
        Enable partitioned processing

        Returns:
            Self for chaining
        """
        self._config["partitioned"] = True
        return self

    def tumbling_time(self, seconds: int) -> "StreamBuilder":
        """
        Configure tumbling time window

        Args:
            seconds: Window duration in seconds

        Returns:
            Self for chaining
        """
        self._config["window_type"] = "tumbling"
        self._config["window_duration_ms"] = seconds * 1000
        return self

    def grace_period(self, seconds: int) -> "StreamBuilder":
        """
        Configure grace period

        Args:
            seconds: Grace period in seconds

        Returns:
            Self for chaining
        """
        self._config["window_grace_period_ms"] = seconds * 1000
        return self

    def lease_timeout(self, seconds: int) -> "StreamBuilder":
        """
        Configure lease timeout

        Args:
            seconds: Lease timeout in seconds

        Returns:
            Self for chaining
        """
        self._config["window_lease_timeout_ms"] = seconds * 1000
        return self

    async def define(self) -> Dict[str, Any]:
        """
        Define/create the stream on the server

        Returns:
            Server response
        """
        return await self._http_client.post("/api/v1/stream/define", self._config)

File: clients/client-js/client-v2/stream/StreamBuilder.js
/**
 * StreamBuilder - Fluent API for defining streams
 */
export class StreamBuilder {
  constructor(httpClient, queen, name, namespace) {
    this.httpClient = httpClient;
    this.queen = queen;
    this.config = {
      name,
      namespace,
      source_queue_names: [],
      partitioned: false,
      window_type: 'tumbling',
      window_duration_ms: 60000,       // 1 minute default
      window_grace_period_ms: 30000,   // 30 seconds default
      window_lease_timeout_ms: 60000   // 1 minute default
    };
  }

  /**
   * Set the source queues for this stream
   * @param {string[]} queueNames - Array of queue names
   */
  sources(queueNames = []) {
    this.config.source_queue_names = queueNames;
    return this;
  }
  
  /**
   * Enable partitioned processing (group by partition_id)
   */
  partitioned() {
    this.config.partitioned = true;
    return this;
  }

  /**
   * Configure tumbling time window
   * @param {number} seconds - Window duration in seconds
   */
  tumblingTime(seconds) {
    this.config.window_type = 'tumbling';
    this.config.window_duration_ms = seconds * 1000;
    return this;
  }
  
  /**
   * Configure grace period for late-arriving messages
   * @param {number} seconds - Grace period in seconds
   */
  gracePeriod(seconds) {
    this.config.window_grace_period_ms = seconds * 1000;
    return this;
  }
  
  /**
   * Configure lease timeout
   * @param {number} seconds - Lease timeout in seconds
   */
  leaseTimeout(seconds) {
    this.config.window_lease_timeout_ms = seconds * 1000;
    return this;
  }

  /**
   * Define/create the stream on the server
   */
  async define() {
    return this.httpClient.post('/api/v1/stream/define', this.config);
  }
}

